utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala

```diff
@@ -0,0 +1,90 @@
+      case (schema: ObjectSchema, map: scala.collection.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ObjectSchema, map: java.util.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ArraySchema, value: Traversable[_]) => encodeCollection(value, schema)
+      case (schema: ArraySchema, value: java.util.Collection[_]) => encodeCollection(value.toArray, schema)
+      case (_: StringSchema, value: String) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Long) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Double) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Float) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Int) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigDecimal) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigInteger) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Number) => Valid(jsonEncoder.encode(value.doubleValue()))
+      // todo dates
+      //      case a: LocalDate => Encoder[LocalDate].apply(a)
+      //      case a: LocalTime => Encoder[LocalTime].apply(a)
+      //      case a: LocalDateTime => Encoder[LocalDateTime].apply(a)
+      //      Default implementation serializes to ISO_ZONED_DATE_TIME which is not handled well by some parsers...
+      //      case a: ZonedDateTime => encodeZonedDateTimeWithFormatter(DateTimeFormatter.ISO_OFFSET_DATE_TIME).apply(a)
+      //      case a: Instant => Encoder[Instant].apply(a)
+      //      case a: OffsetDateTime => Encoder[OffsetDateTime].apply(a)
+      //      case a: UUID => safeString(a.toString)
+      //      case a: DisplayJson => a.asJson
+      case (_: NullSchema, null) => Valid(Json.Null)
+      case (_: NullSchema, None) => Valid(Json.Null)
+      case (_: BooleanSchema, value: Boolean) => Valid(Json.fromBoolean(value))
+      case (_: EnumSchema, value: Enum[_]) => Valid(Json.fromString(value.toString))
+      case (null, value: Any) if validationMode == ValidationMode.lax => Valid(Json.fromString(value.toString))
+      case (_, null) if validationMode != ValidationMode.lax => error(s"Not expected null for field: $fieldName with schema: $schema")
+      case (null, _) if validationMode != ValidationMode.lax => error(s"Not expected null for field: $fieldName with schema: $schema")
+      case (_, _) if validationMode != ValidationMode.lax => error(s"Not expected type: ${value.getClass.getName} for field: $fieldName with schema: $schema")
+    }
+  }
+
+  private def error(str: String): Invalid[NonEmptyList[String]] = Invalid(NonEmptyList.of(str))
+
+  def encode(value: Any, schema: Schema, fieldName: Option[String] = None): WithError[Json] = {
+    optionalEncoders.foldLeft(highPriority)(_.orElse(_)).applyOrElse((value, schema, fieldName), (vs: (Any, Schema, Option[String])) =>
```

Comment:
you can use:
```
case (a: Any, b: Schema, _) =>
```
instead of _1, _2

Reply:
it's not so convenient  because such a partial function would need types

Reply:
My comments is regarding http://twitter.github.io/effectivescala/#Collections-Style . Did you read that? They described many good tips after some wider adoption of scala in twitter. They have many talks about that as well: https://www.youtube.com/results?search_query=twitter+scala

Reply:
refactored